### ISSUE TITLE
Bug of "Use After Free" in coap_io_do_io

### ISSUE THREAD

**Issue by xhzhang824** [obgm/libcoap#775]
I‘m developing a project on Android with libcoap4.3,. There is a crash due to error of "Use After Free, 204 bytes into a 488-byte allocation". The tombstone info is bellow:
```shelll
backtrace:
      #00 pc 000000000002eb44  /system/lib64/libnstackx_ctrl.so (coap_io_do_io+856) (BuildId: be32d415cc27c48f8d9c8693ff9667c9896a2649)

deallocated by thread 29684:
      #00 pc 0000000000047c9c  /apex/com.android.runtime/lib64/bionic/libc.so (scudo::Allocator<scudo::AndroidConfig, &(scudo_malloc_postinit)>::quarantineOrDeallocateChunk(scudo::Options, void*, scudo::Chunk::UnpackedHeader*, unsigned long)+760) (BuildId: fdb81e301ec2fa571b3b9f366f953f3e)
      #01 pc 000000000004369c  /apex/com.android.runtime/lib64/bionic/libc.so (scudo::Allocator<scudo::AndroidConfig, &(scudo_malloc_postinit)>::deallocate(void*, scudo::Chunk::Origin, unsigned long, unsigned long)+212) (BuildId: fdb81e301ec2fa571b3b9f366f953f3e)
      #02 pc 000000000002c32c  /system/lib64/libnstackx_ctrl.so (coap_free_type+32) (BuildId: be32d415cc27c48f8d9c8693ff9667c9896a2649)
      #03 pc 0000000000027838  /system/lib64/libnstackx_ctrl.so (coap_session_free+480) (BuildId: be32d415cc27c48f8d9c8693ff9667c9896a2649)
      #04 pc 0000000000027630  /system/lib64/libnstackx_ctrl.so (coap_session_release+64) (BuildId: be32d415cc27c48f8d9c8693ff9667c9896a2649)
      #05 pc 000000000002eb40  /system/lib64/libnstackx_ctrl.so (coap_io_do_io+852) (BuildId: be32d415cc27c48f8d9c8693ff9667c9896a2649)
      ...[sorry, I cannot show more caller here for information security requirement]

allocated by thread 29684:
      #00 pc 00000000000434d8  /apex/com.android.runtime/lib64/bionic/libc.so (scudo::Allocator<scudo::AndroidConfig, &(scudo_malloc_postinit)>::allocate(unsigned long, scudo::Chunk::Origin, unsigned long, bool)+1308) (BuildId: fdb81e301ec2fa571b3b9f366f953f3e)
      #01 pc 000000000004385c  /apex/com.android.runtime/lib64/bionic/libc.so (scudo_malloc+36) (BuildId: fdb81e301ec2fa571b3b9f366f953f3e)
      #02 pc 000000000003dfbc  /apex/com.android.runtime/lib64/bionic/libc.so (malloc+76) (BuildId: fdb81e301ec2fa571b3b9f366f953f3e)
      #03 pc 000000000002c298  /system/lib64/libnstackx_ctrl.so (coap_malloc_type+32) (BuildId: be32d415cc27c48f8d9c8693ff9667c9896a2649)
      #04 pc 00000000000293c4  /system/lib64/libnstackx_ctrl.so (BuildId: be32d415cc27c48f8d9c8693ff9667c9896a2649)
      #05 pc 00000000000297dc  /system/lib64/libnstackx_ctrl.so (BuildId: be32d415cc27c48f8d9c8693ff9667c9896a2649)
      #06 pc 00000000000296a0  /system/lib64/libnstackx_ctrl.so (coap_new_client_session+28) (BuildId: be32d415cc27c48f8d9c8693ff9667c9896a2649)
      #07 pc 00000000000155f4  /system/lib64/libnstackx_ctrl.so (CoapGetSession+460) (BuildId: be32d415cc27c48f8d9c8693ff9667c9896a2649)
      #08 pc 000000000001586c  /system/lib64/libnstackx_ctrl.so (CoapGetSessionOnTargetServer+296) (BuildId: be32d415cc27c48f8d9c8693ff9667c9896a2649)
      #09 pc 0000000000015b00  /system/lib64/libnstackx_ctrl.so (CoapSendRequest+304) (BuildId: be32d415cc27c48f8d9c8693ff9667c9896a2649)
      #10 pc 0000000000016e14  /system/lib64/libnstackx_ctrl.so (BuildId: be32d415cc27c48f8d9c8693ff9667c9896a2649)
      #11 pc 0000000000030f28  /system/lib64/libnstackx_ctrl.so (BuildId: be32d415cc27c48f8d9c8693ff9667c9896a2649)
      #12 pc 000000000002f640  /system/lib64/libnstackx_ctrl.so (coap_dispatch+584) (BuildId: be32d415cc27c48f8d9c8693ff9667c9896a2649)
      #13 pc 000000000002f368  /system/lib64/libnstackx_ctrl.so (coap_handle_dgram+132) (BuildId: be32d415cc27c48f8d9c8693ff9667c9896a2649)
      #14 pc 0000000000031a5c  /system/lib64/libnstackx_ctrl.so (BuildId: be32d415cc27c48f8d9c8693ff9667c9896a2649)
      #15 pc 000000000002e93c  /system/lib64/libnstackx_ctrl.so (coap_io_do_io+336) (BuildId: be32d415cc27c48f8d9c8693ff9667c9896a2649)
...
``` 
The client session size is 488 and the , so the direct cause is that accessing session after it freed in function **coap_io_do_io**:
```
   if ((s->sock.flags & COAP_SOCKET_CAN_CONNECT) != 0) {
      /* Make sure the session object is not deleted in one of the callbacks  */
      coap_session_reference(s);
      coap_connect_session(ctx, s, now);
      coap_session_release( s );
    }
    if ((s->sock.flags & COAP_SOCKET_CAN_READ) != 0) {
      /* Make sure the session object is not deleted in one of the callbacks  */
      coap_session_reference(s);
      coap_read_session(ctx, s, now);
      coap_session_release(s);
    }
    if ((s->sock.flags & COAP_SOCKET_CAN_WRITE) != 0) {
      /* Make sure the session object is not deleted in one of the callbacks  */
      coap_session_reference(s);
      coap_write_session(ctx, s, now);
      coap_session_release( s );
    }
```
Then I disassemble the project. The action of **#2eb40** is calling function of **coap_session_release**, and the next action is **#2eb44: ldrh	w8, [x21,#204]**, which is meaning **s->sock.flags**. Let’s look back to the backtrace of the crash, we will find that the session is released in **coap_session_release**, then the next action is **if ((s->sock.flags & COAP_SOCKET_CAN_WRITE) != 0) **, which will access the session again. So the root cause is that in function **coap_io_do_io**, for example, we may release a client session when the session's sock.flag is **COAP_SOCKET_CAN_CONNECT**, however we access the session's memory in the next branch checking of **if ((s->sock.flags & COAP_SOCKET_CAN_READ) != 0) **.

**Comment by mrdeep1** [comment]
This is the same issue that you ran into in Issue #745, for which a fix was created in PR #747 which was post 4.3.0 and is now in the develop branch.  Please confirm that your code works as expected when using the develop branch of the code.

**Comment by xhzhang824** [comment]
Yes, thanks! I got that.
